tests/test_client.py

```diff
@@ -113,115 +174,214 @@ def metadata_client():
     ) as client:
         assert code_count_interceptor.counts == {}
-        assert client.Execute(DummyRequest(input="error_once")).output == "OK"
+        if not aio:
+            assert client.Execute(DummyRequest(input="error_once")).output == "OK"
+        else:
+            result = await client.Execute(DummyRequest(input="error_once"))
+            assert result.output == "OK"
         assert code_count_interceptor.counts == {
             grpc.StatusCode.OK: 1,
             grpc.StatusCode.UNKNOWN: 1,
         }
 
 
-def test_caching():
+@pytest.mark.parametrize("aio", [False, True])
+async def test_caching(aio):
     """Caching calls (not calling the continuation) should work."""
-    caching_interceptor = CachingInterceptor()
+    caching_interceptor = AsyncCachingInterceptor() if aio else CachingInterceptor()
     # Use this to test how many times the continuation is called.
-    code_count_interceptor = CodeCountInterceptor()
+    code_count_interceptor = AsyncCodeCountInterceptor() if aio else CodeCountInterceptor()
     interceptors = [caching_interceptor, code_count_interceptor]
-    with dummy_client(special_cases={}, client_interceptors=interceptors) as client:
+    with dummy_client(
+            special_cases={}, client_interceptors=interceptors, aio_server=aio, aio_client=aio,
+    ) as client:
         assert code_count_interceptor.counts == {}
-        assert client.Execute(DummyRequest(input="hello")).output == "hello"
-        assert code_count_interceptor.counts == {grpc.StatusCode.OK: 1}
-        assert client.Execute(DummyRequest(input="hello")).output == "hello"
-        assert code_count_interceptor.counts == {grpc.StatusCode.OK: 1}
-        assert client.Execute(DummyRequest(input="goodbye")).output == "goodbye"
-        assert code_count_interceptor.counts == {grpc.StatusCode.OK: 2}
+        if not aio:
+            assert client.Execute(DummyRequest(input="hello")).output == "hello"
+            assert code_count_interceptor.counts == {grpc.StatusCode.OK: 1}
+            assert client.Execute(DummyRequest(input="hello")).output == "hello"
+            assert code_count_interceptor.counts == {grpc.StatusCode.OK: 1}
+            assert client.Execute(DummyRequest(input="goodbye")).output == "goodbye"
+            assert code_count_interceptor.counts == {grpc.StatusCode.OK: 2}
+        else:
+            result = await client.Execute(DummyRequest(input="hello"))
+            assert result.output == "hello"
+            assert code_count_interceptor.counts == {grpc.StatusCode.OK: 1}
+            result = await client.Execute(DummyRequest(input="hello"))
+            assert result.output == "hello"
+            assert code_count_interceptor.counts == {grpc.StatusCode.OK: 1}
+            result = await client.Execute(DummyRequest(input="goodbye"))
+            assert result.output == "goodbye"
+            assert code_count_interceptor.counts == {grpc.StatusCode.OK: 2}
```

Comment:
Is this any more transparent? does it force equivalence between the two code paths any more?

```python
        if aio:
            async def make_request(word):
                return await client.Execute(DummyRequest(input=word))
        else:
            async def make_request(word):
                return client.Execute(DummyRequest(input=word))

        assert (await make_request("hello")).output == "hello"
        assert code_count_interceptor.counts == {grpc.StatusCode.OK: 1}
        assert (await make_request("hello")).output == "hello"
        assert code_count_interceptor.counts == {grpc.StatusCode.OK: 1}
        assert (await make_request("goodbye")).output == "goodbye"
        assert code_count_interceptor.counts == {grpc.StatusCode.OK: 2}
```